Add knowledge_distillation_tutorial

AlexandrosChrtn opened the pull request:
Includes a Knowledge distillation tutorial.
From this tutorial one can learn:
- How to modify model classes to extract hidden representations and use them for further calculations
- How to modify regular train loops in pytorch to include additional losses on top of, for example, cross-entropy for classification 
- How to improve the performance of lightweight models, using more complex models as teachers

AlexandrosChrtn commented:
@svekars Hello! I get this when deploying, looks related to recent changes in requirements.txt
```
ERROR: No matching distribution found for scipy==1.11.1 (from -r requirements.txt (line 50))
Error installing pip dependencies
Build was terminated: dependency_installation script returned non-zero exit code: 1
Failing build: Failed to install dependencies
```

svekars commented:
@AlexandrosChrtn if you are referring to the Netlify error, it can be safely ignored for now. GitHub Actions checks are the ones we are looking at.

svekars commented:
Can you please add a better description for this tutorial describing how the PyTorch community benefits from this tutorial?

AlexandrosChrtn commented:
@svekars Added a short description with bullets. Let me know if you need anything else!

albanD commented on beginner_source/knowledge_distillation_tutorial.py:
```diff
@@ -0,0 +1,656 @@
+])
+
+# Data preprocessing for CIFAR-10
+transform_test = transforms.Compose([
```
Why not use the same as train one?

albanD commented on beginner_source/knowledge_distillation_tutorial.py:
Which paper?

albanD reviewed (approved):
Sounds good to me!

albanD reviewed (commented):
Thanks for the update.
That sounds good to me cc @svekars

svekars commented on beginner_source/knowledge_distillation_tutorial.py:
```suggestion
# The regressor will be trainable and ideally will do a better job than our naive cosine loss minimization scheme. Its main job is to match the dimensionality of these feature maps so that we can properly define a loss function between the teacher and the student. Defining such a loss function provides a teaching "path," which is basically a flow to back-propagate gradients that will change the student's weights.
```

svekars commented on beginner_source/knowledge_distillation_tutorial.py:
```suggestion
# Focusing on the output of the convolutional layers right before each classifier, we have the following shapes:
```

svekars commented on beginner_source/knowledge_distillation_tutorial.py:
can we make it a ``.. code-block``?

svekars commented on beginner_source/knowledge_distillation_tutorial.py:
```suggestion
# We have 32 filters for the teacher and 16 filters for the student. We will include a trainable layer that converts the feature map of the student to the shape of the feature map of the teacher.
```

svekars commented on beginner_source/knowledge_distillation_tutorial.py:
```suggestion
# After that, we have to update our train loop again. This time, we extract the regressor output of the student, the feature map of the teacher, we calculate the ``MSE`` on these tensors (they have the exact same shape so it's properly defined) and we back propagate gradients based on that loss, in addition to the regular cross entropy loss of the classification task.
```

svekars commented on beginner_source/knowledge_distillation_tutorial.py:
```suggestion
# It is expected that the final method will work better than ``CosineLoss`` because now we have allowed a trainable layer between the teacher and the student, which gives the student some wiggle room when it comes to learning, rather than pushing the student to copy the teacher's representation.
```

svekars commented on beginner_source/knowledge_distillation_tutorial.py:
```suggestion
# Including the extra network is the idea behind hint-based distillation.
```

svekars commented on beginner_source/knowledge_distillation_tutorial.py:
```suggestion
# None of the methods above increases the number of parameters for the network or inference time, so the performance increase comes at the little cost of calculating gradients during training. In ML applications, we mostly care about inference time because training happens before the model deployment.
```

svekars commented on beginner_source/knowledge_distillation_tutorial.py:
Please a dd a Conclusion section that will list everything the user learned in this tutorial and give pointers to more information.

svekars reviewed (commented):
Thanks so much, @AlexandrosChrtn. I added some editorial suggestions. A couple of things, in addition to that:
- Can you some headings that would better break your tutorial to sections (I made some suggestions in my review)
- Please break lines at 79 symbols where possible so that it's easier to edit in the future.

Let me know if you have any questions.

svekars commented:
Also, make sure to check the html output [here](https://docs-preview.pytorch.org/pytorch/tutorials/2514/beginner/knowledge_distillation_tutorial.html ).

svekars commented on beginner_source/knowledge_distillation_tutorial.py:
```diff
@@ -0,0 +1,665 @@
+""" 
+
+######################################################################
+# Knowledge is a technic that enables transfers knowledge from large, computationally expensive
```
```suggestion
# Knowledge distillation is a technique that enables knowledge transfer from large, computationally expensive
```

AlexandrosChrtn commented:
Thanks everyone for the insightful suggestions! @svekars I think I fixed everything based on your proposals, except for the conclusion part, where it feels repetitive to state everything the user learned, becuase we do so in the beginning, however I can still add a few lines if you think it's best. I added sections and removed the lines about filtering warnings. The only reason they were there at the first place is because I saw them in other tutorials (I speculate people add them because of deprecation warnings due to version mismatches and what not).

svekars commented on beginner_source/knowledge_distillation_tutorial.py:
```diff
@@ -0,0 +1,737 @@
+#
+# * 1 GPU, 4GB of memory
+# * PyTorch v2.0 or later 
+# * CIFAR-10 dataset (downloaded by the script and saved it in a directory called ``/data``)
```
```suggestion
# * CIFAR-10 dataset (downloaded by the script and saved in a directory called ``/data``)
```

svekars commented on beginner_source/knowledge_distillation_tutorial.py:
```diff
@@ -0,0 +1,737 @@
+# One function is called ``train`` and takes the following arguments:
+#
+# - ``model``: A model instance to train (update its weights) via this function.
+# - ``train_loader``: we defined our ``train_loader`` above, and its job is to feed the data into the model.
```
```suggestion
# - ``train_loader``: We defined our ``train_loader`` above, and its job is to feed the data into the model.
```

svekars commented on beginner_source/knowledge_distillation_tutorial.py:
```diff
@@ -0,0 +1,737 @@
+# .. figure:: /../_static/img/knowledge_distillation/ce_only.png 
+#    :align: center
+#    
+#    Train both networks with Cross-Entropy. The student will be used as a baseline
```
```suggestion
#    Train both networks with Cross-Entropy. The student will be used as a baseline:
```

svekars commented on beginner_source/knowledge_distillation_tutorial.py:
```diff
@@ -0,0 +1,737 @@
+print(f"LightNN parameters: {total_params_light}")
+
+######################################################################
+# Train and test the lightweight network with cross entropy loss
```
```suggestion
# Train and test the lightweight network with cross entropy loss:
```

svekars commented on beginner_source/knowledge_distillation_tutorial.py:
```diff
@@ -0,0 +1,737 @@
+print("Norm of 1st layer of new_nn_light:", torch.norm(new_nn_light.features[0].weight).item())
+
+######################################################################
+# Print the total number of parameters in each model.
```
```suggestion
# Print the total number of parameters in each model:
```

svekars commented on beginner_source/knowledge_distillation_tutorial.py:
```suggestion
#    Distillation loss is calculated from the logits of the networks. It only returns gradients to the student:
```

svekars commented on beginner_source/knowledge_distillation_tutorial.py:
```suggestion
#    In Cosine Loss minimization, we want to maximize the cosine similarity of the two representations by returning gradients to the student:
```

svekars commented on beginner_source/knowledge_distillation_tutorial.py:
```suggestion
#    The trainable layer matches the shapes of the intermediate tensors and Mean Squared Error (MSE) is properly defined:
```

svekars commented on beginner_source/knowledge_distillation_tutorial.py:
```suggestion
# Conclusion
```

svekars commented on beginner_source/knowledge_distillation_tutorial.py:
```suggestion
# For more information, see:
#
```

svekars reviewed (approved):
Final cleanup
